### DISCUSSION TITLE
ImageRequest method type

### DISCUSSION THREAD

**Discussion by jpappdesigns** [coil-kt/coil#985] [Q&A] [answered]
Hi,

Is there a way to change the method type from GET to POST for ImageRequest? 

Thanks,

JP

**Comment by colinrtwhite** [comment, marked as answer]
Not out of the box, though you can create an extension for this. Here's the code to do this (note: this only works in the 2.x alpha versions):

```kotlin
private const val KEY = "http_method_type"

fun ImageRequest.Builder.httpMethod(method: String) = setParameter(KEY, method)

// Add this interceptor to the OkHttpClient that's used by your ImageLoader instance.
// https://coil-kt.github.io/coil/recipes/#using-a-custom-okhttpclient
class HttpMethodTypeInterceptor : Interceptor {

    fun intercept(chain: Interceptor.Chain): Response {
        val httpMethodType = chain.request.tag(Parameters::class.java)?.value(KEY)
        return if (httpMethodType == null) {
            chain.proceed(chain.request)
        } else {
            val newRequest = chain.request.newBuilder()
                .method(httpMethodType, "".toRequestBody())
                .build()
            chain.proceed(newRequest)
        }
    }
}

// How to use:
imageView.load(url) {
    httpMethod("POST")
}
```

**Reply by jpappdesigns** [reply]
Is it possible to also use this with Compose and rememberImagePainter by passing it in the builder?

**Reply by jpappdesigns** [reply]
Below is the code I have. Each item it runs it throws - Error(painter=BitmapPainter(image=androidx.compose.ui.graphics.AndroidImageBitmap@48729b0, srcOffset=(0, 0), srcSize=175 x 175), result=coil.request.ErrorResult@e918b754).

Each error result has a different value.

Thanks for you help and for my lack of understanding.

`val iLoader = ImageLoader.Builder(context)`
       ` .okHttpClient {`
            `OkHttpClient.Builder()`
               ` .dispatcher(Dispatcher().apply { maxRequestsPerHost = maxRequests })`
               ` .addNetworkInterceptor(HttpMethodTypeInterceptor())`
                `.build()`
       ` }`
       ` .build()`

`val painter = rememberImagePainter(
       data = drawable,
        imageLoader = iLoader,
        builder = {
            placeholder(com.casp.ala.R.drawable.no_image)
            error(com.casp.ala.R.drawable.airnav_icon)
            headers(
                Headers.Builder().set("Authorization", credential)
                    .set("Content-Type", "application/json")
                    .set("Accept", "*/*").build()
            )
            httpMethod("POST")
        }
    )`

**Reply by colinrtwhite** [reply]
Yes, use `httpMethod("POST")` when building your request in `rememberImagePainter`. Also I'd recommend creating your `ImageLoader` in your application to share resources: https://coil-kt.github.io/coil/getting_started/#singleton

**Comment by danielsacco** [comment]
Hi people,

Two years later, I'm using Coil 2.4.0 and this discussion helped me a lot, I needed to POST data to create a QR-Code and show the result on screen.
In this version, the Parameters seem to be lost when HttpUriFetcher creates the okHttp3 Request. Only the tags are mapped into it (line 190), so I had to modify the code a little bit for using tags.

### The interceptor code and extensions:
``` kotlin
data class Body(val content: String)

// Incomplete list of http methods, I'm only using Get/Post
sealed class HttpMethod {
    object Post: HttpMethod()
    object Get: HttpMethod()
}

fun ImageRequest.Builder.httpMethod(method: HttpMethod) = tag(HttpMethod::class.java, method)
fun ImageRequest.Builder.body(body: String) = tag(Body(content = body))

class HttpMethodTypeInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val originalRequest = chain.request()

        // overriden body or the original one
        val body =
            originalRequest.tag(Body::class.java)?.content?.toRequestBody() ?:  
            originalRequest.body

        val request = when(originalRequest.tag(HttpMethod::class.java)) {
            HttpMethod.Get -> originalRequest.newBuilder().method("GET", body).build()
            HttpMethod.Post -> originalRequest.newBuilder() .method("POST", body ?: "".toRequestBody()).build()
            null -> originalRequest
        }

        return chain.proceed(request)
    }
}
```
### In the AsyncImage Code
``` kotlin
        model = ImageRequest.Builder(LocalContext.current)
            .httpMethod(HttpMethod.Post)
            .body("content=$qrContent") 
            .data("https://api.example.com/qr-code")
            ...
            .build(),
        ...
```
### In the Application (As described in [getting_started/image-loaders](https://coil-kt.github.io/coil/getting_started/#image-loaders))
``` kotlin
class MyApplication : Application(), ImageLoaderFactory {
    ...
    override fun newImageLoader(): ImageLoader {
        return ImageLoader.Builder(this)
            .okHttpClient{
                OkHttpClient.Builder()
                    .addInterceptor(HttpMethodTypeInterceptor())
                    .build()
            }
            .build()
    }
```

It worked for me but if you see that I'm doing something stupid please let me know.

Hope this help anybody :)

**Comment by Shyky** [comment]
not working!
